src/API.jl

```diff
@@ -917,9 +925,21 @@ function precompile(ctx::Context)
             break
         end
         if stale
-            Base.compilecache(pkg, sourcepath)
+            t = @async begin
+                if length(pkg_dep_lists[i]) > 0 # if pkg has deps
+                    while length(precomp_list) == 0 || !all(map(x->in(x, precomp_list), pkg_dep_lists[i])) # if all deps not precomped
+                        sleep(0.001)
+                    end
+                end
+                Base.compilecache(pkg, sourcepath)
```

Comment:
To limit parallelization, I suggest launching a million tasks like this, but then creating a `Channel(num_tasks)`, then having each task `put!()` something into it just before this call to `compilecache`, then when it's finished, you `take!()` something back out.  This will create, essentially, an `N`-parallel critical section, and allow `N` tasks to be running that section at once, while all others are blocked, waiting for the channel to free up space.

Reply:
Nice. Ok

Reply:
Would a `Base.Semaphore` be more idiomatic?
Conceptual it is for this, right?

It would be thrrad-safe also.
Which might introduce some overhead?
Idk how hard it is to instantiate a `Semaphore` with a nonthreadsafe `Condition`

Reply:
What happens if this throws?

Reply:
@staticfloat aren't you describing a (counting) `Semaphore`?

Reply:
Yep, pretty much.

Reply:
This PR now has a Semafore approach, and I tried out a channel-based approach here, which doesn't seem simpler https://github.com/JuliaLang/Pkg.jl/compare/master...ianshmean:ib/parallel_precomp_chanelbased

What should we move forward with?

Reply:
cc @tkf just to bring the conversation to a single thread

Reply:
I'd actually implement `precompile` differently by using a recursive function that returns `was_recompiled::Bool` and wrapping the recursive calls with tasks. This way, we don't need to implement a future-like construct (i.e., `was_processed` + `was_recompiled`). The error handling would probably be more straightforward this way. Resource control is probably still easier with semaphore (unless we have an easy-to-use task pool and future in Base or stdlib) although I wish there were `Base.acquire(f, semaphore)`.

But this is the kind of thing the trade-off is not completely clear until you have a concrete implementation. So, I think it's reasonable to defer this to future refactoring.